indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,266 @@
+      @JsonProperty("interval") final Interval interval,
+      @JsonProperty("tuningConfig") final IndexTuningConfig tuningConfig,
+      @JsonProperty("context") final Map<String, Object> context,
+      @JacksonInject Injector injector,
+      @JacksonInject IndexIO indexIO,
+      @JacksonInject ObjectMapper jsonMapper
+  )
+  {
+    super(getOrMakeId(id, TYPE, dataSource), null, taskResource, dataSource, context);
+    this.interval = Preconditions.checkNotNull(interval, "interval");
+    this.tuningConfig = tuningConfig;
+    this.injector = injector;
+    this.indexIO = indexIO;
+    this.jsonMapper = jsonMapper;
+  }
+
+  @JsonProperty
+  public Interval getInterval()
+  {
+    return interval;
+  }
+
+  @JsonProperty
+  public IndexTuningConfig getTuningConfig()
+  {
+    return tuningConfig;
+  }
+
+  @Override
+  public String getType()
+  {
+    return TYPE;
+  }
+
+  @Override
+  public int getPriority()
+  {
+    return getContextValue(Tasks.PRIORITY_KEY, Tasks.DEFAULT_MERGE_TASK_PRIORITY);
+  }
+
+  @Override
+  public boolean isReady(TaskActionClient taskActionClient) throws Exception
+  {
+    final SortedSet<Interval> intervals = new TreeSet<>(Comparators.intervalsByStartThenEnd());
+    intervals.add(interval);
+    return IndexTask.isReady(taskActionClient, intervals);
+  }
+
+  @Override
+  public TaskStatus run(final TaskToolbox toolbox) throws Exception
+  {
+    if (indexTaskSpec == null) {
+      final IndexIngestionSpec ingestionSpec = createIngestionSchema(
+          toolbox,
+          getDataSource(),
+          interval,
+          tuningConfig,
+          indexIO,
+          injector,
+          jsonMapper
+      );
+
+      indexTaskSpec = new IndexTask(
+          getId(),
+          getGroupId(),
+          getTaskResource(),
+          getDataSource(),
+          ingestionSpec,
+          getContext()
+      );
+    }
+
+    final String json = jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(indexTaskSpec);
+    log.info("Generated compaction task details: " + json);
+
+    return indexTaskSpec.run(toolbox);
+  }
+
+  private static IndexIngestionSpec createIngestionSchema(
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval,
+      IndexTuningConfig tuningConfig,
+      IndexIO indexIO,
+      Injector injector,
+      ObjectMapper jsonMapper
+  ) throws IOException, SegmentLoadingException
+  {
+    Pair<Map<DataSegment, File>, List<TimelineObjectHolder<String, DataSegment>>> pair = prepareSegments(
+        toolbox,
+        dataSource,
+        interval
+    );
+    final Map<DataSegment, File> segmentFileMap = pair.lhs;
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = pair.rhs;
+    final List<String> dimensions = IngestSegmentFirehoseFactory.getUniqueDimensions(
+        timelineSegments,
+        new NoopInputRowParser(null)
+    );
+    final List<String> metrics = IngestSegmentFirehoseFactory.getUniqueMetrics(timelineSegments);
```

Comment:
I don't think this is necessary either -- same reason. IngestSegmentFirehoseFactory should include all metrics if null is passed in.

Reply:
Removed.

Reply:
Similar here. CompactionTask finds the unique set of aggregators. I added this line again to make sure the aggregators in the generated DataSchema is used in IngestSegmentFirehose.